Game.render() failure.

**Version:** 0.15.0
**Build date:** 2017-02-05 00:04
**OS:** Windows 7 (x86)
**JRE:** 1.8.0_121
**OpenGL version:** 3.1.0 - Build 9.17.10.4229 (Intel)
**Stack trace:**
```
org.newdawn.slick.SlickException: Game.render() failure.
	at org.newdawn.slick.GameContainer.updateAndRender(GameContainer.java:730)
	at itdelatrisu.opsu.Container.gameLoop(Container.java:135)
	at itdelatrisu.opsu.Container.start(Container.java:79)
	at itdelatrisu.opsu.Opsu.main(Opsu.java:216)
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(Unknown Source)
	at java.util.ArrayList$Itr.next(Unknown Source)
	at itdelatrisu.opsu.ui.NotificationManager.drawNotifications(NotificationManager.java:270)
	at itdelatrisu.opsu.ui.NotificationManager.draw(NotificationManager.java:260)
	at itdelatrisu.opsu.ui.UI.draw(UI.java:124)
	at itdelatrisu.opsu.states.DownloadsMenu.render(DownloadsMenu.java:537)
	at org.newdawn.slick.state.StateBasedGame.render(StateBasedGame.java:199)
	at org.newdawn.slick.GameContainer.updateAndRender(GameContainer.java:727)
	... 3 more
```

--- a/src/itdelatrisu/opsu/ui/NotificationManager.java
+++ b/src/itdelatrisu/opsu/ui/NotificationManager.java
@@ -320,9 +320,9 @@ public class NotificationManager {
 					allFinished = false;
 				startFinished = n.isStartAnimationFinished();
 			}
+			if (allFinished)
+				notifications.clear();  // clear when all are finished showing
 		}
-		if (allFinished)
-			notifications.clear();  // clear when all are finished showing
 
 		// update bar notification
 		if (barNotifTimer > -1 && barNotifTimer < BAR_NOTIFICATION_TIME) {

--- a/src/itdelatrisu/opsu/ui/NotificationManager.java
+++ b/src/itdelatrisu/opsu/ui/NotificationManager.java
@@ -233,10 +233,12 @@ public class NotificationManager {
 			public void mousePressed(int button, int x, int y) {
 				if (button == Input.MOUSE_MIDDLE_BUTTON)
 					return;
-				for (BubbleNotification n : notifications) {
-					if (n.contains(x, y)) {
-						n.click(button == Input.MOUSE_LEFT_BUTTON);
-						break;
+				synchronized (notifications) {
+					for (BubbleNotification n : notifications) {
+						if (n.contains(x, y)) {
+							n.click(button == Input.MOUSE_LEFT_BUTTON);
+							break;
+						}
 					}
 				}
 			}
@@ -267,9 +269,11 @@ public class NotificationManager {
 	 */
 	private void drawNotifications(Graphics g) {
 		int mouseX = input.getMouseX(), mouseY = input.getMouseY();
-		for (BubbleNotification n : notifications) {
-			if (!n.isFinished())
-				n.draw(g, n.contains(mouseX, mouseY));
+		synchronized (notifications) {
+			for (BubbleNotification n : notifications) {
+				if (!n.isFinished())
+					n.draw(g, n.contains(mouseX, mouseY));
+			}
 		}
 	}
 
@@ -307,13 +311,15 @@ public class NotificationManager {
 	public void update(int delta) {
 		// update notifications
 		boolean allFinished = true, startFinished = true;
-		for (BubbleNotification n : notifications) {
-			if (startFinished) {
-				if (n.update(delta))
+		synchronized (notifications) {
+			for (BubbleNotification n : notifications) {
+				if (startFinished) {
+					if (n.update(delta))
+						allFinished = false;
+				} else if (!n.isFinished())
 					allFinished = false;
-			} else if (!n.isFinished())
-				allFinished = false;
-			startFinished = n.isStartAnimationFinished();
+				startFinished = n.isStartAnimationFinished();
+			}
 		}
 		if (allFinished)
 			notifications.clear();  // clear when all are finished showing

